### PR TITLE
Current state of Metafix implementation

### PR DESCRIPTION
**Pull request by fsteeg**
As discussed in our [recent meeting](http://etherpad.lobid.org/p/fix-alma), we want to merge the current state of the Metafix implementation into the main branch to make clear where the main development is happening.

Would it make sense to resolve https://github.com/metafacture/metafacture-fix/issues/35 with this and open new issues for specific improvements / new features?

@blackwinter Feel free to merge yourself as I'll be on vacation starting tomorrow (back on the 25th).

### CODE REVIEW

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    private void processBind(final Do theDo, final EList<String> params) {
+        if (theDo.getName().equals("list")) { // TODO impl multiple binds via FixBind enum
+            final Map<String, String> options = options(theDo.getOptions());
+            final Map<String, Object> fullRecord = new LinkedHashMap<>(record);
+            final Object values = FixMethod.find(record, FixMethod.split(options.get("path")));
+            Metafix.asList(values).stream().filter(val -> val != null).forEach(val -> {
+                // for each val, bind the current record/scope/context to the given var name:
+                record = new LinkedHashMap<>(ImmutableMap.of(options.get("var"), val));
```
Comment (blackwinter) [inline org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java:101]: This prevents access to the actual record. Is this intentional?

File: org.metafacture.fix/build.gradle [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
-  testCompile "org.eclipse.xtext:org.eclipse.xtext.testing:${versions.xtext}"
-  testCompile "org.eclipse.xtext:org.eclipse.xtext.xbase.testing:${versions.xtext}"
+  testImplementation "org.junit.jupiter:junit-jupiter-api:${versions.junit_jupiter}"
+  testImplementation "org.junit.platform:junit-platform-launcher:${versions.junit_platform}"
+  testImplementation "org.eclipse.xtext:org.eclipse.xtext.testing:${versions.xtext}"
+  testImplementation "org.eclipse.xtext:org.eclipse.xtext.xbase.testing:${versions.xtext}"
 
-  testRuntime "org.junit.jupiter:junit-jupiter-engine:${versions.junit_jupiter}"
+  testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:${versions.junit_jupiter}"
+  testRuntimeOnly "org.slf4j:slf4j-simple:${versions.slf4j}"
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/build.gradle]: Done in https://github.com/metafacture/metafacture-fix/pull/60/commits/5a3bd03463f2dd4b630816bd4e457270d9960a24

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+            record.remove(params.get(0));
+            insert(InsertMode.REPLACE, record, split(params.get(0)), params.get(1));
+        }
+    },
+    set_array {
+        public void apply(final Map<String, Object> record, final List<String> params,
+                final Map<String, String> options) {
+            final String key = params.get(0);
+            final List<String> toAdd = params.subList(1, params.size());
+            if (key.contains(APPEND)) {
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java]: Fixed in https://github.com/metafacture/metafacture-fix/pull/60/commits/43d13253072eb2de824c0f4387860e58d345aceb

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+            }
+        }
+    },
+    set_hash {
+        @SuppressWarnings("unchecked")
+        public void apply(final Map<String, Object> record, final List<String> params,
+                final Map<String, String> options) {
+            final String key = params.get(0);
+            final Object val = record.get(key.replace(APPEND, EMPTY));
+            if (key.contains(APPEND) && val instanceof List) {
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java]: Fixed in https://github.com/metafacture/metafacture-fix/pull/60/commits/43d13253072eb2de824c0f4387860e58d345aceb

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static void processList(final InsertMode mode, final String value, final String[] remainingKeys,
+            final Object nested) {
+        final List<Object> nestedList = (List<Object>) nested;
+        final Map<String, Object> nestedMap;
+        switch (remainingKeys[0]) {
+            case "$append":
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java]: Done in https://github.com/metafacture/metafacture-fix/pull/60/commits/111e554ede1839f759b37ef6bfb9090c2152ef20

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+                nestedMap = new LinkedHashMap<>();
+                nestedList.add(nestedMap);
+                insert(mode, nestedMap, Arrays.copyOfRange(remainingKeys, 1, remainingKeys.length), value);
+                break;
+            case "$last":
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java]: Done in https://github.com/metafacture/metafacture-fix/pull/60/commits/111e554ede1839f759b37ef6bfb9090c2152ef20

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+        if (value != null) {
+            final List<Object> vs = Metafix.asList(value);
+            for (final Object v : vs.stream().filter(v -> v != null).collect(Collectors.toList())) {
+                insert(InsertMode.APPEND, record, split(newName), v.toString());
+            }
+        }
+    }
+
+    static String[] split(final String s) {
+        return s.split("\\.");
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:420]: Is this for consistency with Metamorph only or do you have a specific use case?

Since Catmandu uses dot notation, and dot notation is very common, I'd prefer to keep it simple here.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixStandaloneSetup.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    public static void doSetup() {
+        new FixStandaloneSetup().createInjectorAndDoEMFRegistration();
+    }
+
+    public static Fix parseFix(final Reader fixDef) {
+        // TODO: do this only once per application
+        final Injector injector = new FixStandaloneSetup().createInjectorAndDoEMFRegistration();
+        FixStandaloneSetup.doSetup();
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixStandaloneSetup.java]: Merged in 25adea6f8a12b1d8646c18c1765177361c77923a

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:420]: Yes, it's mostly about consistency for now. I haven't had a specific need for this so far, but as soon as you have data with periods in their fields you're stuck (SISIS comes to mind, although we probably won't have to deal with those anymore).

I'm totally fine with postponing this; that's why I created the ticket. We can discuss there if it's worth the effort.

(More or less everything I noted/suggested/questioned here is kind of optional; what doesn't get addressed immediately could/should be turned into tickets for further discussion and prioritization.)

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixStandaloneSetup.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+        try {
+            final URI uri = URI.createFileURI(absPathToTempFile(fixDef, ".fix"));
+            final Resource resource = injector.getInstance(XtextResourceSet.class).getResource(uri, true);
+            final IResourceValidator validator = ((XtextResource) resource).getResourceServiceProvider().getResourceValidator();
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixStandaloneSetup.java]: Merged in 25adea6f8a12b1d8646c18c1765177361c77923a

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:420]: > what doesn't get addressed immediately could/should be turned into tickets for further discussion and prioritization

:+1:

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+    }
+
+    private void emit(final Object key, final Object val) {
+        if (val == null) {
+            return;
+        }
+        final List<?> vals = asList(val);
+        final boolean isMulti = vals.size() > 1 || val instanceof List;
+        if (isMulti) {
+            outputStreamReceiver.startEntity(key.toString() + "[]");
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:155]: Added in https://github.com/metafacture/metafacture-fix/issues/59#issuecomment-958931286

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+            if (value instanceof Map) {
+                final Map<?, ?> nested = (Map<?, ?>) value;
+                outputStreamReceiver.startEntity(isMulti ? "" : key.toString());
+                nested.entrySet().forEach(nestedEntry -> {
+                    emit(nestedEntry.getKey(), nestedEntry.getValue());
+                });
+                outputStreamReceiver.endEntity();
+            }
+            else {
+                outputStreamReceiver.literal(isMulti ? (i + 1) + "" : key.toString(), value.toString());
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:168]: Added in https://github.com/metafacture/metafacture-fix/issues/59#issuecomment-958931286

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java [hunk 12]
```diff
@@ -30,161 +30,255 @@
...
     private StreamReceiver streamReceiver;
 
-    public MetafixIfTest() {
+    public MetafixBindTest() {
     }
 
     @Test
-    public void ifTopLevel() {
+    public void doList() {
         MetafixTestHelpers.assertFix(streamReceiver, Arrays.asList(//
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java]: Removed in https://github.com/metafacture/metafacture-fix/pull/60/commits/15977a4e4581399ad542a7ab0295bf5ee34b3418

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java [hunk 13]
```diff
@@ -30,161 +30,255 @@
...
-                "map('name')",
-                "if contains('name', 'University')", //
-                "  add_field('type', 'Organization')", //
-                "end"), //
+                "do list('path': 'name', 'var': 'n')",
+                " upcase('n')",
+                " trim('n')",
+                " copy_field('n', 'author')",
+                "end",
+                "remove_field('name')"), //
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java]: Fine with me, but should be part of the Checkstyle definition (contains similar check for if/else).

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java [hunk 14]
```diff
@@ -30,161 +30,255 @@
...
             i -> {
                 i.startRecord("1");
-                i.literal("name", "A University");
+                i.literal("name", " A University");
+                i.literal("name", "Max ");
                 i.endRecord();
             }, o -> {
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java:64]: Fine with me, but should be part of the Checkstyle definition (contains similar check for if/else).

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java [hunk 15]
```diff
@@ -30,161 +30,255 @@
...
                 o.get().literal("name", "Max");
+                o.get().literal("type", "Person");
+                f.apply(2).endEntity();
                 o.get().endRecord();
-                o.get().startRecord("2");
+            });
+    }
+
+    @Test
+    @Disabled // TODO: how to handle repeated entities: turn to array vs. merge because it's the same?
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java]: Done in https://github.com/metafacture/metafacture-fix/pull/60/commits/4f06e3ae2b3a9ba359c63a9f039bbb468c3d37db

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 16]
```diff
@@ -89,10 +97,71 @@ public void apply(final Multimap<String, String> record, final List<String> para
...
+                }
+            });
+        }
 
+        @SuppressWarnings("unchecked")
+        private Map<String, Integer> getNamedGroups(final Pattern regex) {
+            try {
+                // Not available as API, see https://stackoverflow.com/a/15596145/18154:
+                final Method namedGroupsMethod = Pattern.class.getDeclaredMethod("namedGroups");
+                namedGroupsMethod.setAccessible(true);
```
Comment (blackwinter) [inline org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: This doesn't seem to work when consumed as an external dependency (Maven package or Gradle source dependency; internal tests work fine):

```
java.lang.reflect.InaccessibleObjectException: Unable to make java.util.Map java.util.regex.Pattern.namedGroups() accessible: module java.base does not "opens java.util.regex" to unnamed module @30ee2816
    at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354) ~[?:?]
    at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297) ~[?:?]
    at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199) ~[?:?]
    at java.base/java.lang.reflect.Method.setAccessible(Method.java:193) ~[?:?]
    at org.metafacture.metafix.FixMethod$11.getNamedGroups(FixMethod.java:168) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.FixMethod$11.lambda$apply$1(FixMethod.java:146) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at java.base/java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
    at org.metafacture.metafix.FixMethod$11.apply(FixMethod.java:141) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.RecordTransformer.processFunction(RecordTransformer.java:170) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.RecordTransformer.processSubexpressions(RecordTransformer.java:89) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.RecordTransformer.transform(RecordTransformer.java:64) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.metafix.Metafix.endRecord(Metafix.java:137) ~[metafacture-fix-0.2.0-SNAPSHOT.jar:?]
    at org.metafacture.flowcontrol.StreamBuffer.replay(StreamBuffer.java:64) ~[metafacture-flowcontrol-5.2.0.jar:?]
    at org.metafacture.metamorph.Filter.dispatch(Filter.java:84) ~[metamorph-5.2.0.jar:?]
    at org.metafacture.metamorph.Filter.endRecord(Filter.java:99) ~[metamorph-5.2.0.jar:?]
[...]
```

Can you confirm? If so, we'd have to find a different solution.

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java [hunk 17]
```diff
@@ -0,0 +1,703 @@
...
+import java.util.Arrays;
+
+/**
+ * Tests Metafix record level methods. Following the cheat sheet
+ * examples at https://github.com/LibreCat/Catmandu/wiki/Fixes-Cheat-Sheet
+ *
+ * @author Fabian Steeg
+ */
+@ExtendWith(MockitoExtension.class)
+@SuppressWarnings("checkstyle:MultipleStringLiterals")
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java:39]: Suggesting switching to `@SuppressWarnings` in https://github.com/metafacture/metafacture-fix/pull/60/commits/ea54a0ec96a2bb90a47f057f70c05c83d563fd82, as it's the standard Java suppression mechanism.

[the remark this answers is not in the document]
Comment (blackwinter) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java:39]: But it's also limited in where it can appear! Now the check is suppressed in the whole method `org.metafacture.metafix.InterpreterTest.shouldInterpretNested()`; should we - unintentionally - add another "magic number", we wouldn't notice.

Also:

1. At least in our case, check violations are errors, not warnings.
2. We can't limit the number of allowed exceptions (currently two).
3. The `SuppressWithNearbyCommentFilter` should be removed if we're no longer using it.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java (hunk 16 above)
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: It <a href="https://test.metafacture.org/playground/?data=1%7Bdate%3A+2015-03-07%7D&flux=as-lines%0A%7Cdecode-formeta%0A%7Cfix%0A%7Cencode-json&fix=parse_text%28date%2C+%27%28%3F%3Cyear%3E\\d%7B4%7D%29-%28%3F%3Cmonth%3E\\d%7B2%7D%29-%28%3F%3Cday%3E\\d%7B2%7D%29%27%29&morph=%3C%3Fxml+version%3D%221.0%22+encoding%3D%22UTF-8%22%3F%3E%0A%3Cmetamorph+xmlns%3D%22http%3A//www.culturegraph.org/metamorph%22+xmlns%3Axsi%3D%22http%3A//www.w3.org/2001/XMLSchema-instance%22%0A%09version%3D%221%22%3E%0A%09%3Crules%3E%0A%09%09%3Cdata+source%3D%22_id%22+name%3D%22id%22/%3E%0A%09%09%3Cdata+source%3D%22a%22+name%3D%22title%22/%3E%0A%09%09%3Ccombine+value%3D%22%24%7Bfirst%7D+%24%7Blast%7D+aus+%24%7Bplace%7D%22+name%3D%22author%22%3E%0A%09%09%09%3Cdata+source%3D%22b.v%22+name%3D%22first%22+/%3E%0A%09%09%09%3Cdata+source%3D%22b.n%22+name%3D%22last%22+/%3E%0A%09%09%09%3Cdata+source%3D%22c%22+name%3D%22place%22+/%3E%0A%09%09%3C/combine%3E%0A%09%3C/rules%3E%0A%3C/metamorph%3E%0A&active-editor=fix">works in the playground</a>, which consumes metafacture-fix via the local Maven repo, but that's probably because the playground also runs on Java 8, so we will have to find a solution that works with the Java module system.

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: > but that's probably because the playground also runs on Java 8

Yeah, sorry, should have mentioned that. Limetrans is on 17, which puts the onus on me, I guess... :frowning_face:

**Comment by fsteeg** [comment]
> Why are the tests creating `List<String>`s? Can't we just use string concatenation as in metafacture-core?

I wanted to avoid having to add `\n` all over the samples (in metamorph we don't need them).

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java (hunk 16 above)
Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: Would you be open to adding a simple guard for now, so we can at least use `parse_text` with numbered groups?

```java
final Map<String, Integer> namedGroups = params.get(1).contains("(?<") ? getNamedGroups(p) : Collections.emptyMap();
```

Then we can postpone the search for a compatible solution.

Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: Yes, good idea!

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: I _might_ have an idea for a solution. Stay tuned.

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: See 5395e90 for a possible alternative. Do you think the requirements are reasonable?

Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: > 1. Named groups are not escaped/quoted

Not sure what you mean here. Do you mean it won't work to match a literal `(?<some>thing)` (by escaping it)?

> 2. Named groups are not mixed with unnamed groups.

Yes, I think it makes sense to either use named or unnamed groups, not both.

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: > Not sure what you mean here. Do you mean it won't work to match a literal (?<some>thing) (by escaping it)?

Yes. If the pattern contains a character sequence that matches `(?<.+?>` but isn't actually a named group (e.g., `\\(?<name>` - parenthesis is escaped - or `\Q(?<.+?>\E` - pattern is quoted), then it won't work as intended.

> Yes, I think it makes sense to either use named or unnamed groups, not both.

There's no meaningful output to generate anyway (and this wasn't supported before either), but now it would also lead to wrong output as the group counts may be off. E.g., `(?<a>.)(.)(?<c>.)` with input `abc` produces `{"a":"a","c":"b"}`, while previously it was `{"a":"a","c":"c"}`

Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:168]: Right, yes I think these are both reasonable, +1 for https://github.com/metafacture/metafacture-fix/commit/5395e907ee69a9d61f3a5dcb705d30cfa7b2e840, thanks for working on that!

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java]: > as the group counts may be off.

Wait a minute, why do we even use the group index when we've got the name! Fixed in 26d2a9e (plus tests for these edge cases).

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (blackwinter) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java]: I have no idea how to configure Checkstyle to that effect. Let's ignore it for now.

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (blackwinter) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixBindTest.java:64]: I have no idea how to configure Checkstyle to that effect. Let's ignore it for now.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/validation/XtextValidator.java [hunk 18]
```diff
@@ -29,12 +33,10 @@ public static boolean validate(final URI uri) throws IOException {
         final int count = issues.size();
 
         if (count > 0) {
-            System.out.println(String.format("The Xtext file '%s' has %d issue%s:\n",
-                        resource.getURI(), count, count > 1 ? "s" : ""));
+            LOG.warn("The Xtext file '{}' has {} issue{}:", resource.getURI(), count, count > 1 ? "s" : "");
```
[the remark this answers is not in the document]
Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/validation/XtextValidator.java]: Fixed by @fsteeg in 5a3bd03.

**Comment by fsteeg** [comment]
> Why were `ServerLauncher` and `FixServlet` removed in [a786f73](https://github.com/metafacture/metafacture-fix/commit/a786f73a4a654578902c6f4ee1cad9456dc32bfa)?

Hm, that must have happened accidentally during refactoring, I restored them in https://github.com/metafacture/metafacture-fix/pull/60/commits/e4dec3aef20ed025150f43cea4fe7cba582f6a0d.

> Why was `MetafixFilter` removed in [f328a13](https://github.com/metafacture/metafacture-fix/commit/f328a1360bbb4e576133db9dd1741a47591b234a)? [...] commit without any mention of the removal

Right, that should have been a separate commit, but was intentionally. My idea was that we don't need a separate filter command with fix, since we can filter in the main fix (or in a separate fix, but with normal fix behavior), [e.g. like this](https://github.com/metafacture/metafacture-fix/blob/e4dec3aef20ed025150f43cea4fe7cba582f6a0d/org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java#L611).

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java:39]: Opened https://github.com/metafacture/metafacture-fix/issues/62 to follow up on this.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+            throw new IllegalStateException(ENTITIES_NOT_BALANCED);
+        }
+        flattener.endRecord();
+        LOG.debug("End record, walking fix: {}", currentRecord);
+        final RecordTransformer transformer = new RecordTransformer(currentRecord, vars, fix);
+        currentRecord = transformer.transform();
+        if (!currentRecord.containsKey("__reject")) {
+            outputStreamReceiver.startRecord(recordIdentifier);
+            LOG.debug("Sending results to {}", outputStreamReceiver);
+            currentRecord.keySet().stream().filter(k -> !k.startsWith("_")).forEach(k -> {
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:141]: Opened https://github.com/metafacture/metafacture-fix/issues/63 to follow up on this.

**Comment by fsteeg** [comment]
> I would suggest to create a `Record` class that wraps the `Map<String, Object>` instead of "littering" `Metafix` and `FixMethod` with static methods.

Yes, good point. I opened https://github.com/metafacture/metafacture-fix/issues/64 for that.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java (hunk 1 above)
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java:101]: My idea was that in the scope of the bind, the record (as the subexpressions see it) should only contain the bound value. I'm not sure why though, that might very well make no sense.

**Comment by blackwinter** [comment]
> My idea was that we don't need a separate filter command with fix, since we can filter in the main fix (or in a separate fix, but with normal fix behavior)

Agreed, `reject()` is a viable alternative.

File: org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java [hunk 20]
```diff
@@ -0,0 +1,703 @@
...
+                i.startEntity("nested");
+                i.literal("key", "val2");
+                i.endEntity();
+                i.endEntity();
+                i.endRecord();
+            }, (o, f) -> {
+                o.get().startRecord("1");
+                o.get().startEntity("deep");
+                o.get().startEntity("nested");
+                o.get().startEntity("key[]");
```
[the remark this answers is not in the document]
Comment (fsteeg) [reply org.metafacture.fix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java:134]: Yeah, the whole topic of repeated values / arrays needs work. I've opened https://github.com/metafacture/metafacture-fix/issues/65 for that.

**Comment by fsteeg** [comment]
@blackwinter Alright, I hope I have addressed or opened issues for all your points. Reassigning & re-requesting review.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java (hunk 1 above)
Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java:101]: I wasn't able to find it documented anywhere, but Catmandu seems to preserve access to the whole record:

```perl
$fixes = <<EOF;
do list(path:foo,var:loop)
  copy_field(test,loop.baz)
end
EOF

$fixer = Catmandu::Fix->new(fixes => [$fixes]);

is_deeply $fixer->fix({foo => [{bar => 1}, {bar => 2}], test => 42}),
    {foo => [{bar => 1, baz => 42}, {bar => 2, baz => 42}], test => 42},
    'binding scope testing';
```

**Comment by blackwinter** [comment]
> I hope I have addressed or opened issues for all your points.

Yes, I believe everything's been addressed. Thank you! :smile: 

Once the conversation regarding the binding scope in `do list` has been resolved (either by opening an issue or, preferably, by changing the implementation), this pull request can be merged. I'll hold off on the final approval until then.

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java (hunk 1 above)
Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java:101]: I agree, the full record should be available. I've opened https://github.com/metafacture/metafacture-fix/issues/66 since this does require some work that I probably won't have time to do this week (simply using the full record with the current implementation results in overwriting previous values under the same key, which is covered by a test).

**Review by blackwinter — APPROVED** [review]
OK, let's move on then :)

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+        if (name == null) {
+            throw new IllegalArgumentException("Entity name must not be null.");
+        }
+        ++entityCount;
+        final Integer currentEntityIndex = entityCountStack.peek() - 1;
+        final Map<String, Object> previousEntity = currentEntityIndex < 0 ||
+                entities.size() <= currentEntityIndex ? null : entities.get(currentEntityIndex);
+        entityCountStack.push(Integer.valueOf(entityCount));
+        flattener.startEntity(name);
+        entities.add(currentEntity(name, previousEntity == null && entities.size() >= 0 ? currentRecord : previousEntity));
```
Comment (blackwinter) [inline org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:187]: What's the purpose of `&& entities.size() >= 0`? (`List.size()` can't be negative.)

Or is it a typo? (`> 0`? `|| entities.size() == 0`?)

Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:187]: Right, that makes no sense. I think we don't (no longer?) need that logic there at all.

```suggestion
        entities.add(currentEntity(name, previousEntity != null ? previousEntity : currentRecord));
```

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:187]: Good, I'll (silently) remove it during my refactor for #64 if that's okay with you.

Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/Metafix.java:187]: :+1:

File: org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java [hunk 22]
```diff
@@ -0,0 +1,432 @@
...
+        final String[] remainingKeys = Arrays.copyOfRange(keys, 1, keys.length);
+        return findNested(map, currentKey, remainingKeys);
+    }
+
+    private static Object findNested(final Map<String, Object> map, final String currentKey,
+            final String[] remainingKeys) {
+        final Object nested = map.get(currentKey);
+        // TODO: array of maps, like in insertNested
+        if (nested instanceof List) {
+            return ((List<?>) nested).stream().map(o -> findNested(map, currentKey, remainingKeys))
```
Comment (blackwinter) [inline org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:364]: What is this code supposed to do? (It's not currently covered by any tests.) I read it as: For each element in this list, find this list (at `currentKey`) in the map - thus **recursively** multiplying the list by its size!?! Was it maybe intended to find `remainingKeys` in each element (`o`)? (The comment above seems to indicate as much.)

Comment (fsteeg) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:364]: Yeah, weird, it should be doing something with each element `o`. Since no test fails without it, I think we should remove it and re-implement with a test later, if we need it.

Comment (blackwinter) [reply org.metafacture.fix/src/main/java/org/metafacture/metafix/FixMethod.java:364]: OK, I'll just remove it during my refactoring (#72) in an extra commit.
